<launch>
<!-- Top-Level launch file for bundled launching /-->

	<!-- Available arguments -->
	<arg name="manualMode" default="false"/>
	<arg name="autoMode" default="false"/>
	<arg name="uav_name" default="elev_0"/>
	<arg name="visuals" default="false"/>
	<arg name="log" default="true"/>
    <arg name="shutdownOnScenarioEnd" default="true"/>
    <arg name="fe_name" default="elev_0"/>
    <arg name="log_name" default="16"/>

    <!-- Launch last-letter simulation nodes -->
    <include file="$(find uav_ftc)/launch/last_letter_nodes.launch" >
        <arg name="manualMode" value="$(arg manualMode)"/>
        <arg name="autoMode" value="$(arg autoMode)"/>
        <arg name="uav_name" value="$(arg uav_name)"/>
        <arg name="visuals" value="$(arg visuals)"/>
    </include>
    
    <!-- Launch uav_ftc-related nodes -->
    <group ns="$(arg uav_name)">
        <!-- Load controller parameters -->
        <rosparam>
            ctrlMode: 3
            deltaFMin: 0.0
            pMin: -1.0
            qMin: -1.0
            rMin: -10
            deltaFMax: 12.0
            pMax: 1.0
            qMax: 1.0
            rMax: 10
            referenceMin: [10.0, -0.52, -0.5]
            referenceMax: [20.0,  0.52,  0.5]
            waypointRadius: 20
            numWpLookahead: 5
        </rosparam>

        <!--  -->
        <!-- Launch logger -->
        <!--  -->
        <group if="$(arg log)">
            <include file="$(find uav_ftc)/launch/logger.launch">
                <arg name="custom_name" value="true"/>
                <arg name="log_name" value="$(arg log_name)"/>
            </include>
        </group>

        <!-- Launch the scenario player -->
        <!-- <node pkg="uav_ftc" name="scenario_player" type="scenario_player.py" output="screen" required="$(arg shutdownOnScenarioEnd)">
            <param name="scenario" value="$(arg scenario)"/>
        </node> -->

        <!-- Launch static FE publisher -->
        <node pkg="uav_ftc" name="fe_publisher" type="fe_publisher.py" output="screen">
            <param name="fe" value="$(arg fe_name)"/>
        </node>

        <!-- Launch the databus generator -->
        <rosparam command="load" file="$(find uav_ftc)data/parameters/data_bus.yaml" />
        <node pkg="uav_ftc" name="data_bus" type="data_bus" output="screen">
            <param name="data_source" value="1"/>
        </node>

        <!--  -->
        <!-- Launch controllers -->
        <!--  -->
        
        <!-- Raise controller -->
        <node pkg="uav_ftc" name="controller" type="controller_no_elev" respawn="true" output="screen">
            <param name="~rate" value="50"/>
            <param name="~kp_beta" value="2.0"/>
            <param name="~ki_beta" value="0.0"/>
            <param name="~kd_beta" value="0"/>
            <param name="~kp_gamma" value="1"/>
            <param name="~ki_gamma" value="1.0"/>
            <param name="~kd_gamma" value="0"/>
            <param name="~kp_psi_dot" value="10"/>
            <param name="~ki_psi_dot" value="20"/>
            <param name="~kd_psi_dot" value="0"/>
            <param name="~kp_phi" value="0.2"/>
            <param name="~ki_phi" value="0.1"/>
            <param name="~kd_phi" value="0"/>
            <param name="~kp_p" value="5"/>
            <param name="~ki_p" value="0.1"/>
            <param name="~kd_p" value="0"/>
        </node>

        <!-- Load obstacles -->
        <rosparam command="load" file="$(find uav_planner)data/parameters/map_sparse.yaml" />

        <!-- Raise path controller -->
        <node pkg="uav_ftc" name="path_controller" type="path_controller" respawn="true" output="screen">
            <param name="~rate" value="2"/>
            <remap from="~waypointRadius" to="waypointRadius"/>
            <remap from="~numWpLookahead" to="numWpLookahead"/>
        </node>

        <!-- Playback waypoints broadcasting from log file-->
        <node pkg="rosbag" name="replayer" type="play" args="-d 2 $(find uav_ftc)logs/waypoints_sparse_filtered_synched.bag">
            <remap from="/skywalker_2013_mod/obstacles" to="/$(arg uav_name)/obstacles"/>
            <remap from="/skywalker_2013_mod/rrt_path" to="/$(arg uav_name)/rrt_path"/>
            <remap from="/skywalker_2013_mod/waypoints" to="/$(arg uav_name)/waypoints"/>
        </node>

        <!-- Raise input aggreggator node -->
        <node pkg="uav_ftc" name="input_aggregator" type="input_aggregator" output="screen">
            <remap from="~rate" to="/world/simRate"/>
        </node>

    </group>

</launch>